include/etcpal/cpp/inet.h

```diff
@@ -0,0 +1,836 @@
+/// \ingroup etcpal_cpp
+/// \brief C++ utilities for the \ref etcpal_inet module.
+///
+/// Provides C++ wrappers for IP addresses (IpAddr), socket addresses (SockAddr) and MAC addresses
+/// (MacAddr).
+
+/// \ingroup etcpal_cpp_inet
+/// \brief Indicates an IP address family, or an invalid IP address.
+enum class IpAddrType
+{
+  Invalid = kEtcPalIpTypeInvalid,
+  V4 = kEtcPalIpTypeV4,
+  V6 = kEtcPalIpTypeV6
+};
+
+/// \ingroup etcpal_cpp_inet
+/// \brief A wrapper class for the EtcPal IP address type.
+///
+/// Provides C++ syntactic sugar for working with IP addresses.
+class IpAddr
+{
+public:
+  ETCPAL_CONSTEXPR_14 IpAddr() noexcept;
+  // Note: this constructor is not explicit by design, to allow implicit conversions e.g.
+  //   etcpal::IpAddr ip = etcpal_c_function_that_returns_ip();
+  constexpr IpAddr(const EtcPalIpAddr& c_ip) noexcept;
+  IpAddr& operator=(const EtcPalIpAddr& c_ip) noexcept;
+
+  ETCPAL_CONSTEXPR_14 IpAddr(uint32_t v4_data) noexcept;
+  explicit IpAddr(const uint8_t* v6_data) noexcept;
+  IpAddr(const uint8_t* v6_data, unsigned long scope_id) noexcept;
+
+  constexpr const EtcPalIpAddr& get() const noexcept;
+  ETCPAL_CONSTEXPR_14 EtcPalIpAddr& get() noexcept;
+  std::string ToString() const;
+  constexpr uint32_t v4_data() const noexcept;
+  constexpr const uint8_t* v6_data() const noexcept;
+  std::array<uint8_t, ETCPAL_IPV6_BYTES> ToV6Array() const;
+  constexpr unsigned long scope_id() const noexcept;
+
+  constexpr bool IsValid() const noexcept;
+  constexpr IpAddrType type() const noexcept;
+  constexpr bool IsV4() const noexcept;
+  constexpr bool IsV6() const noexcept;
+  bool IsLinkLocal() const noexcept;
+  bool IsLoopback() const noexcept;
+  bool IsMulticast() const noexcept;
+  bool IsWildcard() const noexcept;
+  unsigned int MaskLength() const noexcept;
+
+  void SetAddress(uint32_t v4_data) noexcept;
+  void SetAddress(const uint8_t* v6_data) noexcept;
+  void SetAddress(const uint8_t* v6_data, unsigned long scope_id) noexcept;
+
+  static IpAddr FromString(const char* ip_str) noexcept;
+  static IpAddr FromString(const std::string& ip_str) noexcept;
+  static IpAddr WildcardV4() noexcept;
+  static IpAddr WildcardV6() noexcept;
+  static IpAddr Wildcard(IpAddrType type) noexcept;
+  static IpAddr NetmaskV4(unsigned int mask_length) noexcept;
+  static IpAddr NetmaskV6(unsigned int mask_length) noexcept;
+  static IpAddr Netmask(IpAddrType type, unsigned int mask_length) noexcept;
+
+private:
+  EtcPalIpAddr addr_{};
+};
+
+/// \brief Constructs an invalid IP address by default.
+ETCPAL_CONSTEXPR_14_OR_INLINE IpAddr::IpAddr() noexcept
+{
+  ETCPAL_IP_SET_INVALID(&addr_);
+}
+
+/// \brief Construct an IP address copied from an instance of the C EtcPalIpAddr type.
+constexpr IpAddr::IpAddr(const EtcPalIpAddr& c_ip) noexcept : addr_(c_ip)
+{
+}
+
+/// \brief Assign an instance of the C EtcPalIpAddr type to an instance of this class.
+inline IpAddr& IpAddr::operator=(const EtcPalIpAddr& c_ip) noexcept
```

Comment:
is there a reason that these two and quite a few below them are marked as constexpr and inline instead of using our ETCPAL_CONSTEXPR_14_OR_INLINE

Reply:
Here is my understanding of the issues that led to me doing it the way I've done it:

The `ETCPAL_CONSTEXPR_14` macro solves the problem of functions that can be defined `constexpr` in C++14 but not in C++11; namely, functions that have multiple statements, or non-const member functions that can still be `constexpr`.

Functions that are defined in a header outside of the class definition must be marked `inline`. `constexpr` implies `inline`.

This results in the following rules:
- If a function doesn't satisfy the requirements of `constexpr` in any C++ standard (e.g. `operator=`, `IpAddr()` constructor containing a call to `memcpy()`, etc), its definition should just be marked `inline`.
- If a function satisfies the requirements of `constexpr` in C++14 but not C++11, its definition should be marked `ETCPAL_CONSTEXPR_14_OR_INLINE`, which makes it `constexpr` in 14 and above, and `inline` in 11.
- If a function satisfies the requirements of `constexpr` in C++11 onward, its definition should just be marked `constexpr`.

Does that make sense?